### DISCUSSION TITLE
Questions about -v variable=value & parsing .arguments()

### DISCUSSION THREAD

**Discussion by cascading-jox** [c4spar/cliffy#318] [Q&A] [answered]
Hi! 

I have some questions: 
1. Is it possible to do have options like -v var=val like in awk? I tried this but to no avail: 
```
.option("-v, --variable <variable-name>=<value>", "pass variable(s) to script")
```
2. What is the correct way to parse arguments passed with .arguments()? Right now I use `.action((_, arg) => inArg = arg)` which feels like a hack. Intuitively I assumed it would be in the returned object from `new Command()`. 
3. How do you check if the argument to the option is valid? For example I have this option but any argument to the option is valid: 
```
.option("-g, --generate <time-series|polar>")
```

**Comment by c4spar** [comment, marked as answer]
Hi @cascading-jox,

sry for late reply, i somehow missed this discussion.

**1.)**

> Is it possible to do have options like -v var=val like in awk? I tried this but to no avail:

There are different ways to achieve your goal:

[**Value handler**](https://cliffy.io/docs@v0.24.2/command/options#map-option-value)

```ts
await new Command()
  .option(
    "-v, --variable <variable-name>",
    "pass variable(s) to script",
    {
      value: (val) => {
        const [key, ...parts] = val.split("=");
        const value = parts.join("=");
        if (!value) {
          throw new ValidationError(`missing value for variable: ${val}`);
        }
        return { [key]: value };
      },
    },
  )
  .action(({ variable }) => console.log("variable:", variable))
  .parse(Deno.args);
```

<details>
  <summary>In combination with the list type:</summary>

```ts
await new Command()
  .option(
    "-v, --variable <variable-name:string[]>",
    "pass variable(s) to script",
    {
      value: (values, prev: Record<string, string> = {}) => {
        for (const val of values) {
          const [key, ...parts] = val.split("=");
          const value = parts.join("=");
          if (!value) {
            throw new ValidationError(`missing value for variable: ${val}`);
          }
          prev[key] = value;
        }
        return prev;
      },
    },
  )
  .action(({ variable }) => console.log("variable:", variable))
  .parse(Deno.args);
```
</details>

<details>
  <summary>In combination with the collect option:</summary>

```ts
await new Command()
  .option(
    "-v, --variable <variable-name>",
    "pass variable(s) to script",
    {
      collect: true,
      value: (val, prev: Record<string, string> = {}) => {
        const [key, ...parts] = val.split("=");
        const value = parts.join("=");
        if (!value) {
          throw new ValidationError(`missing value for variable: ${val}`);
        }
        return { ...prev, [key]: value };
      },
    },
  )
  .action(({ variable }) => console.log("variable:", variable))
  .parse(Deno.args);
```
</details>

[**Custom type**](https://cliffy.io/docs@v0.24.2/command/types#custom-types)

```ts
await new Command()
  .type("variable", ({ label, name, type, value }) => {
    const [key, ...parts] = value.split("=");
    const val = parts.join("=");
    if (!val) {
      throw new ValidationError(
        `${label} "${name}" must be of type "${type}", but got "${value}". Expacted value: <variable>=<value>`,
      );
    }
    return { [key]: val };
  })
  .option(
    "-v, --variable <variable-name:variable>",
    "pass variable(s) to script",
  )
  .action(({ variable }) => console.log("variable:", variable))
  .parse(Deno.args);
```

A custom type can be also combined with the `collect` option and `list` type but it will return an array of objects instead of a single object. But the array could be also mapped in the value handler to merge all objects into a single one.

**2.)**

> What is the correct way to parse arguments passed with .arguments()?

Arguments supports also build-in and custom types: `.arguments("<foo:number> <bar:string> <baz:my-custom-type>")`.

I planned to also add a `.argument()` method to the `Command` class to also add support for value handlers etc. but this is currently not implemented.

**3.)**

> How do you check if the argument to the option is valid?

Types can be defined after the argument name separated by colon: `.option("-g, --generate <name:type>")`

There are a few [build-in](https://cliffy.io/docs@v0.24.2/command/types#build-in-types) types and you can also define your [custom types](https://cliffy.io/docs@v0.24.2/command/types#custom-types).
For sutom validation and value mapping you can use the [value handler](https://cliffy.io/docs@v0.24.2/command/options#map-option-value) to manually validate the value for a specific option:

```ts
new Command()
  .option("-f, --foo <name>", "Description...", ({value}) => {
    // validate value...
    return value;
  })
```

or:


```ts
new Command()
  .option("-f, --foo <name>", "Description...", {
    value: ({value}) => {
      // validate value...
      return value;
    }
  })
```

**Reply by cascading-jox** [reply]
Thank you! Just was I was looking for!
